_posts/on-working-with-viewmodel-composition-based-system.md

```diff
@@ -0,0 +1,145 @@
+
+Let's think about the development phase of service A. It's composed of two components: the restful API mediating the data access. The composition handler allows data to flow from the restful API to the view model owned by branding.
+
+## A repository for each logical service
+
+If service A is developed using .NET, it'll be a class library for the composition handler and a web application for the restful API. They could both be in the same solution and repository. We simplify the development and testing phase with the risk of making the deployment a bit more complex. Let's investigate why.
+
+There is coupling between the composition handler and the web API, but there isn't tight coupling. There might be cases in which one component changes without forcing any other in the logical service to change. If that's the case, it means that every deployment of one component doesn't correspond to the deployment of the others. Or in other words, they can be versioned separately. However, if they reside in the same repository, it'll be harder to have different CI/CD pipelines for the various components. Usually, there is a one-to-one relationship between a repository and CI/CD pipelines.
+We could compromise and decide to release all the components in the same repository whenever we deploy one. Fine, but consider that there will be "wasted" version numbers, or call them empty releases, for all the non-touched elements, which might result in maintenance issues.
+
+To recap, if we're going down the route of a single repository for each logical service, there will be pros and cons:
+
+- Pros: the overall development process will be smoother
+- Pros: multiple components integration testing will be straightforward
+- Cons: more complex CI/CD pipelines
+- Cons: each release will release all components, even though some might be released with no changes
+
+What else can we do?
+
+## A repository per component 
+
+We could have a different repository for each component: A repository for the restful APIs, one for each composition handler, another for the composed ViewModel, and a last one for the frontend.
+
+CI/CD pipelines would be much easier. The restful API can be deployed independently from other components, and the same goes for all the components. Composition handlers could be deployed as Nuget packages, consumed by the composition gateway host at deployment time. Composition handlers and the fronted would also depend on the ViewModel package.
+
+By doing so, we'd get fine-grained control over packages deployment and versioning. It would be much easier to set up and customize CI/CD pipelines for each component. And we won't waste any version number because each component would only be released when strictly necessary.
```

Comment:
```suggestion
By doing so, we'd get fine-grained control over package deployment and versioning. It would be much easier to set up and customize CI/CD pipelines for each component. And we won't waste any version numbers because each component would only be released when strictly necessary.
```